### PR TITLE
recursive api_calls are forbidden

### PR DESCRIPTION
**Pull request by lnussel**
delay stop_job if needed to wait fo update_status

### CODE REVIEW

File: lib/OpenQA/Worker/Jobs.pm [hunk 1]
```diff
@@ -97,6 +99,27 @@ sub stop_job($;$) {
 
     _kill_worker($worker);
 
+    # XXX: we need to wait if there is an update_status in progress.
+    # we should have an event emitter that subscribes to update_status done
+    my $stop_job_check_status;
+    $stop_job_check_status = sub {
+        if($update_status_running) {
```
Comment (aaannz) [inline lib/OpenQA/Worker/Jobs.pm]: Wouldn't this dead lock if stop_job(api-failure) is triggered during api_call?

**Comment by aaannz** [comment]
I can't say I like it. I find it making things even more complicated. 

I agree with the part about saving update_status running status, however I'm currently testing change to one shot timer.
I don't like splitting stop_job however.

By ( now properly) blocking update_status we eliminate regular api_call so next interference may only come from 1) backend is done, 2) web socket command is received. Right?

File: lib/OpenQA/Worker/Jobs.pm (hunk 1 above)
Comment (lnussel) [inline lib/OpenQA/Worker/Jobs.pm]: it won't even go there as $stop_job_running would be set then

**Comment by lnussel** [comment]
noone can seriously like anything in the worker :) It needs a rewrite as proper mojo class with event emitter etc. maybe then there is a chance to become pretty. Another place with timer and api_call is check_job as I just saw.

File: lib/OpenQA/Worker/Jobs.pm (hunk 1 above)
Comment (aaannz) [inline lib/OpenQA/Worker/Jobs.pm]: Right!

File: lib/OpenQA/Worker/Common.pm [hunk 2]
```diff
@@ -144,15 +145,19 @@ sub api_init {
...
     my ($method, $path, $params, $json_data, $ignore_errors) = @_;
+    state $call_running;
 
     return undef unless verify_workerid();
+
     if ($call_running) {
-        print "api call in progress, not accepting new one\n" if $verbose;
-        return undef;
+        # quit immediately
+        Mojo::IOLoop->next_tick(sub {} );
```
Comment (aaannz) [inline lib/OpenQA/Worker/Common.pm]: What is the purpose of this?

Comment (lnussel) [inline lib/OpenQA/Worker/Common.pm]: Mojo::IOLoop->stop only stops on the next event that occurs. so this call introduces one :-)

**Pull request merged**